File: 11/main.py

import math
import numpy as np

def load_input(input_file: str) -> list[int]:
	f = open(input_file, 'r')

	return [int(x) for x in list(f.readlines()[0].strip('\n').split(' '))]

def blink(prev:list[int]):
	stones = []

	for s in prev:
		stones += calc_blink(s)

	return stones

def calc_blink(num:int) -> int:
	if num == 0:
		return [1]
	
	num_digits = 1 + math.floor(math.log10(num))

	if num_digits % 2 == 1:
		return [num * 2024]

	t = math.pow(10, num_digits / 2)

	right = math.floor(num % t)
	left = math.floor((num - right)/t)

	return [right, left]


""" def blink_single(start:int, blinks:int) -> int:
	max_size = 1_000_000_000
	size = 1
	arr1 = np.zeros(max_size)
	arr2 = np.zeros(max_size)

	prev = [start]

	for b in range(blinks):
		for i in range(size):
			
		print(len(prev))
		stones = []
		for s in prev:
			if s == 0:
				stones.append(1)
				continue

			num_digits = 1 + math.floor(math.log10(s))

			if num_digits % 2 == 1:
				stones.append(s * 2024)
				continue

			t = math.pow(10, num_digits / 2)

			right = math.floor(s % t)
			left = math.floor((s - right)/t)

			stones.append(left)
			stones.append(right)

		prev = stones.copy()

	return len(prev)
 """

def main():
	stones = load_input('input.txt')
	total_stones = 0
	num_blinks = 25
	
	print('initial arrangement\n', stones)
	for b in range(num_blinks):
		stones = blink(stones)

	total_stones = len(stones)
	print("total stones\n", total_stones)
	assert(total_stones == 207683)

    
if __name__=='__main__':
	main()

File: 11/main.rs


fn blink(num:u64, i:u64, depth:u64) -> u64 {
    if i == depth {
        return 1
    }

    if num ==  0 {
        return blink(1, i+1, depth);
    }

    let f = num as f64; 
    let num_digits: u32 = 1 + f.log(10.0).floor() as u32;

    if num_digits % 2 == 1 {
        return blink(num*2024, i+1, depth);
    }

    //println!("{}", num_digits);

    let ten: u64 = 10;
    let t: u64 = ten.pow(num_digits/2);

    //println!("{}", t);

    let right = num % t;
    let left = (num - right)/t;

    
    //println!("{}", left);
    //println!("{}", right);

    return blink(left,i+1,depth) + blink(right,i+1, depth);

}

fn main() {
    let depth = 75;

    let result = 
        blink(3935565, 0, depth) + 
        blink(31753, 0, depth) + 
        blink(437818, 0, depth) + 
        blink(7697, 0, depth) + 
        blink(5, 0, depth) +
        blink(38, 0, depth) +
        blink(0, 0, depth) +
        blink(123, 0, depth);

    println!("{}", result);
}

